File: resume_parser.py
"""Resume Parser — parses PDF, DOCX, and TXT resumes."""
import re, os, json
from pathlib import Path

try:
    import fitz  # PyMuPDF
    HAS_PYMUPDF = True
except ImportError:
    HAS_PYMUPDF = False

try:
    import pdfplumber
    HAS_PDFPLUMBER = True
except ImportError:
    HAS_PDFPLUMBER = False

try:
    import docx
    HAS_DOCX = True
except ImportError:
    HAS_DOCX = False

SKILLS_DB = {
    "programming_languages": [
        "python", "java", "javascript", "typescript", "c", "c++", "c#", "go", "rust",
        "ruby", "php", "swift", "kotlin", "scala", "r", "matlab", "perl", "lua",
        "haskell", "elixir", "erlang", "dart", "flutter", "bash", "shell", "powershell",
        "groovy", "objective-c", "assembly", "cobol", "fortran", "julia", "solidity",
    ],
    "frameworks_libraries": [
        "react", "angular", "vue.js", "node.js", "django", "flask", "spring boot",
        "express.js", "rails", "laravel", "fastapi", "nestjs", "next.js", "nuxt.js",
        "gatsby", "svelte", "tensorflow", "pytorch", "keras", "scikit-learn",
        "pandas", "numpy", "matplotlib", "seaborn", "plotly", "opencv",
        "bootstrap", "tailwind", "jquery", "redux", "graphql", "rest api",
        "microservices", "docker", "kubernetes", "aws", "azure", "gcp",
        "terraform", "ansible", "jenkins", "git", "github", "gitlab", "bitbucket",
        "spring", "hibernate", "mybatis", "struts", "hadoop", "spark", "kafka",
        "rabbitmq", "celery", "airflow", "dbt", "langchain", "huggingface",
    ],
    "databases": [
        "mongodb", "mysql", "postgresql", "sqlite", "oracle", "sql server", "redis",
        "elasticsearch", "cassandra", "dynamodb", "firebase", "supabase",
        "mariadb", "neo4j", "influxdb", "cockroachdb", "snowflake", "bigquery",
    ],
    "soft_skills": [
        "leadership", "communication", "teamwork", "problem solving", "critical thinking",
        "creativity", "adaptability", "time management", "project management",
        "decision making", "negotiation", "presentation skills", "analytical thinking",
    ],
    "tools_technologies": [
        "git", "docker", "kubernetes", "aws", "azure", "gcp", "linux", "unix",
        "vscode", "intellij", "pycharm", "eclipse", "visual studio",
        "postman", "jira", "confluence", "figma", "tableau", "power bi",
        "grafana", "prometheus", "kibana", "splunk", "datadog",
        "ci/cd", "devops", "agile", "scrum", "kanban",
    ],
    "domains": [
        "machine learning", "deep learning", "artificial intelligence", "data science",
        "data analysis", "data engineering", "nlp", "computer vision",
        "web development", "mobile development", "cloud computing",
        "cybersecurity", "blockchain", "iot", "full stack", "frontend", "backend",
        "api development", "system design", "distributed systems",
        "fintech", "healthcare", "e-commerce", "gaming",
    ],
}

STOP_WORDS = {
    "the", "is", "at", "which", "and", "are", "was", "were", "be", "been", "being",
    "have", "has", "had", "do", "does", "did", "will", "would", "could", "should",
    "may", "might", "must", "shall", "can", "to", "of", "in", "for", "on", "with",
    "about", "against", "between", "into", "through", "during", "before", "after",
    "above", "below", "from", "up", "down", "out", "off", "over", "under", "again",
    "further", "then", "once", "here", "there", "when", "where", "why", "how", "all",
    "both", "each", "few", "more", "most", "other", "some", "such", "no", "nor", "not",
    "only", "own", "same", "so", "than", "too", "very", "s", "t", "just", "don", "now",
    "also", "about", "an", "as", "but", "by", "he", "her", "his", "i", "if", "it", "me",
    "my", "myself", "or", "she", "them", "their", "theirs", "these", "they", "this",
    "those", "we", "what", "which", "who", "whom", "you", "your", "yours",
}


class ResumeParser:
    """Parses resumes in PDF, DOCX, and TXT formats to extract structured data."""

    def __init__(self):
        self.skills_database = SKILLS_DB

    def parse_resume(self, file_path: str) -> dict:
        """Parse a resume file and extract structured data."""
        ext = os.path.splitext(file_path)[1].lower()
        text = self._extract_text(file_path, ext)
        return self._extract_data(text)

    def _extract_text(self, file_path: str, ext: str) -> str:
        """Extract raw text from a resume file."""
        if ext == ".pdf":
            return self._parse_pdf(file_path)
        elif ext == ".docx":
            return self._parse_docx(file_path)
        elif ext == ".txt":
            return self._parse_txt(file_path)
        else:
            raise ValueError(f"Unsupported file format: {ext}")

    def _parse_pdf(self, file_path: str) -> str:
        """Parse PDF resume."""
        if HAS_PYMUPDF:
            doc = fitz.open(file_path)
            text = ""
            for page in doc:
                text += page.get_text() + "\n"
            doc.close()
            return text
        elif HAS_PDFPLUMBER:
            with pdfplumber.open(file_path) as pdf:
                return "\n".join(page.extract_text() or "" for page in pdf.pages)
        else:
            raise ImportError(
                "No PDF parsing library available. Install one:\n"
                "  pip install pymupdf        # recommended\n"
                "  pip install pdfplumber     # alternative"
            )

    def _parse_docx(self, file_path: str) -> str:
        """Parse DOCX resume."""
        if not HAS_DOCX:
            raise ImportError("python-docx not installed. Install with: pip install python-docx")
        doc = docx.Document(file_path)
        text = "\n".join([para.text for para in doc.paragraphs])
        return text

    def _parse_txt(self, file_path: str) -> str:
        """Parse TXT resume."""
        with open(file_path, "r", encoding="utf-8") as f:
            return f.read()

    def _extract_data(self, text: str) -> dict:
        """Extract structured data from resume text."""
        return {
            "name": self._extract_name(text),
            "email": self._extract_email(text),
            "phone": self._extract_phone(text),
            "location": self._extract_location(text),
            "skills": self._extract_skills(text),
            "experience": self._extract_experience(text),
            "education": self._extract_education(text),
            "projects": self._extract_projects(text),
            "certifications": self._extract_certifications(text),
            "languages": self._extract_languages(text),
            "summary": self._extract_summary(text),
            "keywords": self._extract_keywords(text),
            "experience_years": self._calculate_experience_years(text),
        }

    def _extract_name(self, text: str) -> str:
        """Extract name from the first lines of the resume."""
        lines = text.split("\n")[:5]
        for line in lines:
            trimmed = line.strip()
            if trimmed and len(trimmed) < 50 and not re.search(r"[@\d]", trimmed):
                if re.match(r"^[A-Za-z\s.'-]+$", trimmed):
                    return trimmed
        return "Unknown"

    def _extract_email(self, text: str) -> str:
        """Extract email address."""
        match = re.search(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b", text)
        return match.group(0) if match else ""

    def _extract_phone(self, text: str) -> str:
        """Extract phone number."""
        match = re.search(r"\+?\d[\d\s\-().]{9,14}\d", text)
        return match.group(0).strip() if match else ""

    def _extract_location(self, text: str) -> str:
        """Extract location/city."""
        patterns = [
            r"(?:Location|Address|City|Based in)\s*[:.]?\s*([A-Za-z\s,]+)",
            r"([A-Z][a-z]+(?:[,\s]+[A-Z][a-z]+)*),?\s*(?:India|USA|UK|Canada|Australia|Germany)",
        ]
        for pattern in patterns:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                return match.group(1).strip()[:60]
        return ""

    def _extract_skills(self, text: str) -> list:
        """Extract skills from the resume using the skills database."""
        text_lower = text.lower()
        found_skills = set()

        for skill in self.skills_database.values():
            for skill_name in skill:
                sl = skill_name.lower()
                if sl not in found_skills:
                    pattern = r"(?<![a-z])" + re.escape(sl) + r"(?![a-z])"
                    if re.search(pattern, text_lower):
                        found_skills.add(sl)

        return list(found_skills)

    def _extract_section(self, text: str, *section_names: str) -> str:
        """Extract a section from the resume by section heading."""
        pattern = r"(?:" + "|".join(re.escape(n) for n in section_names) + r")\s*\n([\s\S]*?)(?=\n\s*(?:Experience|Work Experience|Education|Skills|Projects|Certifications|Languages|Interests|Hobbies|References|Achievements)\s*\n|$)"
        match = re.search(pattern, text, re.IGNORECASE)
        return match.group(1) if match else ""

    def _split_entries(self, section_text: str) -> list:
        """Split a section into individual entries."""
        entries = re.split(r"\n\s*\n|\n\s*[-•*]\s*|\n\s*\d+\.\s*", section_text)
        return [e.strip() for e in entries if len(e.strip()) > 15][:5]

    def _extract_experience(self, text: str) -> list:
        """Extract work experience entries."""
        return self._split_entries(self._extract_section(
            text, "Experience", "Work Experience", "Professional Experience", "Employment"
        ))

    def _extract_education(self, text: str) -> list:
        """Extract education entries."""
        return self._split_entries(self._extract_section(
            text, "Education", "Academic Background", "Qualifications"
        ))

    def _extract_projects(self, text: str) -> list:
        """Extract project entries."""
        return self._split_entries(self._extract_section(
            text, "Projects", "Personal Projects", "Key Projects", "Portfolio"
        ))

    def _extract_certifications(self, text: str) -> list:
        """Extract certification entries."""
        return self._split_entries(self._extract_section(
            text, "Certifications", "Certificates", "Licenses"
        ))

    def _extract_languages(self, text: str) -> list:
        """Extract spoken languages."""
        section = self._extract_section(text, "Languages", "Language Proficiency")
        if not section:
            return []
        languages = [p.strip() for p in re.split(r"[,\n•*]+", section) if len(p.strip()) > 3]
        return languages[:5]

    def _extract_summary(self, text: str) -> str:
        """Extract professional summary."""
        section = self._extract_section(
            text, "Summary", "Professional Summary", "Profile", "About Me", "Objective", "Career Objective"
        )
        if not section:
            return ""
        cleaned = re.sub(r"(?:Professional Summary|Summary|About Me|Profile|Career Objective|Objective)\s*[:.]?\s*", "", section, flags=re.IGNORECASE)
        return cleaned.strip()[:500] if len(cleaned.strip()) > 20 else ""

    def _extract_keywords(self, text: str) -> list:
        """Extract top keywords from the resume."""
        tokens = re.findall(r"\b[a-z]+\b", text.lower())
        filtered = [w for w in tokens if w not in STOP_WORDS and len(w) > 3]
        freq = {}
        for w in filtered:
            freq[w] = freq.get(w, 0) + 1
        return [word for word, _ in sorted(freq.items(), key=lambda x: x[1], reverse=True)[:20]]

    def _calculate_experience_years(self, text: str) -> int:
        """Estimate years of experience."""
        # Look for explicit mentions
        patterns = [
            r"(\d+)\s*\+?\s*(?:years?|yrs?)\s*(?:of\s*)?(?:experience|exp)",
            r"(?:experience|exp)\s*[:\-]?\s*(\d+)\s*(?:years?|yrs?)",
        ]
        for pattern in patterns:
            match = re.search(pattern, text, re.IGNORECASE)
            if match:
                return int(match.group(1))

        # Fallback: count year ranges
        years = re.findall(r"\b(20\d{2}|19\d{2})\b", text)
        if len(years) >= 2:
            year_nums = [int(y) for y in years]
            span = max(year_nums) - min(year_nums)
            if 0 < span <= 40:
                return span

        return 0


if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1:
        parser = ResumeParser()
        result = parser.parse_resume(sys.argv[1])
        print(json.dumps(result, indent=2))
    else:
        print("Usage: python resume_parser.py <resume_file>")

File: static/resume_parser.js
/**
 * Client-side Resume Parser
 * Parses PDF, DOCX, and TXT resumes entirely in the browser
 * Uses pdf.js for PDF parsing and mammoth.js for DOCX parsing
 */

// Skills database (same as server-side)
const SKILLS_DATABASE = {
    programming_languages: [
        'python', 'java', 'javascript', 'typescript', 'c', 'c++', 'c#', 'go', 'rust',
        'ruby', 'php', 'swift', 'kotlin', 'scala', 'r', 'matlab', 'perl', 'lua',
        'haskell', 'elixir', 'erlang', 'dart', 'flutter', 'bash', 'shell', 'powershell',
        'groovy', 'objective-c', 'assembly', 'cobol', 'fortran', 'julia', 'solidity',
    ],
    frameworks_libraries: [
        'react', 'angular', 'vue.js', 'node.js', 'django', 'flask', 'spring boot',
        'express.js', 'rails', 'laravel', 'fastapi', 'nestjs', 'next.js', 'nuxt.js',
        'gatsby', 'svelte', 'tensorflow', 'pytorch', 'keras', 'scikit-learn',
        'pandas', 'numpy', 'matplotlib', 'seaborn', 'plotly', 'opencv',
        'bootstrap', 'tailwind', 'jquery', 'redux', 'graphql', 'rest api',
        'microservices', 'docker', 'kubernetes', 'aws', 'azure', 'gcp',
        'terraform', 'ansible', 'jenkins', 'git', 'github', 'gitlab', 'bitbucket',
        'spring', 'hibernate', 'mybatis', 'struts', 'hadoop', 'spark', 'kafka',
        'rabbitmq', 'celery', 'airflow', 'dbt', 'langchain', 'huggingface',
    ],
    databases: [
        'mongodb', 'mysql', 'postgresql', 'sqlite', 'oracle', 'sql server', 'redis',
        'elasticsearch', 'cassandra', 'dynamodb', 'firebase', 'supabase',
        'mariadb', 'neo4j', 'influxdb', 'cockroachdb', 'snowflake', 'bigquery',
    ],
    soft_skills: [
        'leadership', 'communication', 'teamwork', 'problem solving', 'critical thinking',
        'creativity', 'adaptability', 'time management', 'project management',
        'decision making', 'negotiation', 'presentation skills', 'analytical thinking',
    ],
    tools_technologies: [
        'git', 'docker', 'kubernetes', 'aws', 'azure', 'gcp', 'linux', 'unix',
        'vscode', 'intellij', 'pycharm', 'eclipse', 'visual studio',
        'postman', 'jira', 'confluence', 'figma', 'tableau', 'power bi',
        'grafana', 'prometheus', 'kibana', 'splunk', 'datadog',
        'ci/cd', 'devops', 'agile', 'scrum', 'kanban',
    ],
    domains: [
        'machine learning', 'deep learning', 'artificial intelligence', 'data science',
        'data analysis', 'data engineering', 'nlp', 'computer vision',
        'web development', 'mobile development', 'cloud computing',
        'cybersecurity', 'blockchain', 'iot', 'full stack', 'frontend', 'backend',
        'api development', 'system design', 'distributed systems',
        'fintech', 'healthcare', 'e-commerce', 'gaming',
    ],
};

// Stop words for keyword extraction
const STOP_WORDS = new Set([
    'the', 'is', 'at', 'which', 'and', 'are', 'was', 'were', 'be', 'been', 'being',
    'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could', 'should',
    'may', 'might', 'must', 'shall', 'can', 'to', 'of', 'in', 'for', 'on', 'with',
    'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after',
    'above', 'below', 'from', 'up', 'down', 'out', 'off', 'over', 'under', 'again',
    'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all',
    'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not',
    'only', 'own', 'same', 'so', 'than', 'too', 'very', 's', 't', 'just', 'don', 'now',
    'also', 'about', 'an', 'as', 'but', 'by', 'he', 'her', 'his', 'i', 'if', 'it', 'me',
    'my', 'myself', 'or', 'she', 'them', 'their', 'theirs', 'these', 'they', 'this',
    'those', 'we', 'what', 'which', 'who', 'whom', 'you', 'your', 'yours', 'and', 'or',
    'nor', 'but', 'yet', 'so', 'a', 'an', 'the', 'am', 'is', 'are', 'was', 'were', 'be',
    'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a',
    'an', 'the', 'will', 'just', 'don', 'should', 'now', 'i', 'me', 'my', 'myself', 'we',
    'our', 'ours', 'ourselves', 'you', 'your', 'yours', 'yourself', 'yourselves', 'he',
    'him', 'his', 'himself', 'she', 'her', 'hers', 'herself', 'it', 'its', 'itself',
    'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom',
    'this', 'that', 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been',
    'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an',
    'the', 'may', 'must', 'shall', 'can', 'ought', 'will', 'would', 'should', 'could',
    'might', 'i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', 'your',
    'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', 'her', 'hers',
    'herself', 'it', 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves',
    'what', 'which', 'who', 'whom', 'this', 'that', 'these', 'those', 'am', 'is', 'are',
    'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does',
    'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until',
    'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into',
    'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down',
    'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here',
    'there', 'when', 'where', 'why', 'how', 'all', 'both', 'each', 'few', 'more', 'most',
    'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', 'too',
    'very', 's', 't', 'can', 'will', 'just', 'don', 'now', 'also', 'about', 'with', 'from',
    'for', 'by', 'in', 'on', 'at', 'to', 'and', 'or', 'but', 'is', 'are', 'was', 'were',
    'be', 'been', 'being', 'have', 'has', 'had', 'has', 'had', 'have', 'has', 'had',
    'having', 'do', 'does', 'did', 'doing', 'will', 'would', 'could', 'should', 'may',
    'might', 'must', 'shall', 'can', 'cannot', 'cant', 'ain', 'isn', 'aren', 'wasn',
    'weren', 'hasn', 'haven', 'hadn', 'doesn', 'don', 'didn', 'won', 'wouldn', 'couldn',
    'shouldn', 'mustn', 'shan', 'needn', 'daren', 'oughtn', 'need', 'used', 'dare',
    'need', 'ought', 'used', 'dare', 'need', 'used', 'dare', 'need', 'used', 'dare',
    'need', 'used', 'dare', 'need', 'used', 'dare', 'need', 'used', 'dare', 'need',
    'used', 'dare', 'need', 'used', 'dare', 'need', 'used', 'dare', 'need', 'used',
    'dare', 'need', 'used', 'dare', 'need', 'used', 'dare', 'need', 'used', 'dare',
    'need', 'used', 'dare', 'need', 'used', 'dare', 'need', 'used', 'dare', 'need',
    'used', 'dare', 'need', 'used', 'dare', 'need', 'used', 'dare', 'need', 'used',
]);

class ResumeParser {
    constructor() {
        this.skillsDatabase = SKILLS_DATABASE;
    }

    /**
     * Parse a resume file (PDF, DOCX, or TXT)
     * @param {File} file - The resume file
     * @returns {Promise<Object>} Parsed resume data
     */
    async parseResume(file) {
        const ext = file.name.split('.').pop().toLowerCase();
        let text = '';

        try {
            if (ext === 'pdf') {
                text = await this._parsePDF(file);
            } else if (ext === 'docx') {
                text = await this._parseDOCX(file);
            } else if (ext === 'txt') {
                text = await this._parseTXT(file);
            } else {
                throw new Error(`Unsupported file format: .${ext}`);
            }
        } catch (err) {
            console.error('PDF parsing error:', err);
            throw new Error('Failed to parse resume. Please try a different file format.');
        }

        return {
            name: this._extractName(text),
            email: this._extractEmail(text),
            phone: this._extractPhone(text),
            location: this._extractLocation(text),
            skills: this._extractSkills(text),
            experience: this._extractExperience(text),
            education: this._extractEducation(text),
            projects: this._extractProjects(text),
            certifications: this._extractCertifications(text),
            languages: this._extractLanguages(text),
            summary: this._extractSummary(text),
            keywords: this._extractKeywords(text),
            experience_years: this._calculateExperienceYears(text),
        };
    }

    async _parsePDF(file) {
        const arrayBuffer = await file.arrayBuffer();
        const pdf = await pdfjsLib.getDocument({ data: arrayBuffer }).promise;
        let text = '';
        for (let i = 1; i <= pdf.numPages; i++) {
            const page = await pdf.getPage(i);
            const content = await page.getTextContent();
            text += content.items.map(item => item.str).join(' ') + '\n';
        }
        return text;
    }

    async _parseDOCX(file) {
        const arrayBuffer = await file.arrayBuffer();
        const result = await mammoth.extractRawText({ arrayBuffer });
        return result.value;
    }

    async _parseTXT(file) {
        return await file.text();
    }

    _extractName(text) {
        const lines = text.split('\n').slice(0, 5);
        for (const line of lines) {
            const trimmed = line.trim();
            if (trimmed && trimmed.length < 50 && !/[@\d]/.test(trimmed) && /^[A-Za-z]/.test(trimmed)) {
                return trimmed;
            }
        }
        return 'Unknown';
    }

    _extractEmail(text) {
        const match = text.match(/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/);
        return match ? match[0] : '';
    }

    _extractPhone(text) {
        const match = text.match(/\+?\d[\d\s\-().]{9,14}\d/);
        return match ? match[0].trim() : '';
    }

    _extractLocation(text) {
        const patterns = [
            /(?:Location|Address|City|Based in)\s*[:.]?\s*([A-Za-z\s,]+)/i,
            /([A-Z][a-z]+(?:[,\s]+[A-Z][a-z]+)*),?\s*(?:India|USA|UK|Canada|Australia|Germany)/i,
        ];
        for (const pattern of patterns) {
            const match = text.match(pattern);
            if (match) {
                return match[1].trim().substring(0, 60);
            }
        }
        return '';
    }

    _extractSkills(text) {
        const textLower = text.toLowerCase();
        const found = new Set();
        for (const skills of Object.values(this.skillsDatabase)) {
            for (const skill of skills) {
                const sl = skill.toLowerCase();
                if (!found.has(sl) && new RegExp(`(?<![a-z])${this._escapeRegex(sl)}(?![a-z])`).test(textLower)) {
                    found.add(sl);
                }
            }
        }
        return Array.from(found);
    }

    _escapeRegex(str) {
        return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    }

    _section(text, ...headings) {
        const pattern = `(?:${headings.map(h => this._escapeRegex(h)).join('|')})[\\s\\S]*?(?=\\n\\s*(?:experience|work experience|education|skills|projects|certifications|languages|interests|hobbies|references|achievements)\\s*\\n|$)`;
        const match = text.match(new RegExp(pattern, 'i'));
        return match ? match[0] : '';
    }

    _splitEntries(sectionText) {
        return sectionText
            .split(/\n\s*\n|\n\s*[-•*]\s*|\n\s*\d+\.\s*/)
            .map(e => e.trim())
            .filter(e => e.length > 15)
            .slice(0, 5);
    }

    _extractExperience(text) {
        return this._splitEntries(this._section(text, 'Experience', 'Work Experience', 'Professional Experience', 'Employment'));
    }

    _extractEducation(text) {
        return this._splitEntries(this._section(text, 'Education', 'Academic Background', 'Qualifications'));
    }

    _extractProjects(text) {
        return this._splitEntries(this._section(text, 'Projects', 'Personal Projects', 'Key Projects', 'Portfolio'));
    }

    _extractCertifications(text) {
        return this._splitEntries(this._section(text, 'Certifications', 'Certificates', 'Licenses'));
    }

    _extractLanguages(text) {
        const section = this._section(text, 'Languages', 'Language Proficiency');
        return section
            .split(/[,\n•*]+/)
            .map(p => p.trim())
            .filter(p => p.length > 3 && p.length < 30)
            .slice(0, 5);
    }

    _extractSummary(text) {
        const section = this._section(text, 'Summary', 'Professional Summary', 'Profile', 'About Me', 'Objective', 'Career Objective');
        const cleaned = section.replace(/(?:Professional Summary|Summary|About Me|Profile|Career Objective|Objective)\s*[:.]?\s*/gi, '');
        return cleaned.trim().length > 20 ? cleaned.trim().substring(0, 500) : '';
    }

    _extractKeywords(text) {
        const tokens = text.toLowerCase().match(/\b[a-z]+\b/g) || [];
        const filtered = tokens.filter(w => !STOP_WORDS.has(w) && w.length > 3);
        const freq = {};
        filtered.forEach(w => { freq[w] = (freq[w] || 0) + 1; });
        return Object.entries(freq)
            .sort((a, b) => b[1] - a[1])
            .slice(0, 20)
            .map(([word]) => word);
    }

    _calculateExperienceYears(text) {
        const patterns = [
            /(\d+)\s*\+?\s*(?:years?|yrs?)\s*(?:of\s*)?(?:experience|exp)/i,
            /(?:experience|exp)\s*[:\-]?\s*(\d+)\s*(?:years?|yrs?)/i,
        ];
        for (const pattern of patterns) {
            const match = text.match(pattern);
            if (match) return parseInt(match[1], 10);
        }

        const years = text.match(/\b(20\d{2}|19\d{2})\b/g);
        if (years && years.length >= 2) {
            const yearNums = years.map(y => parseInt(y, 10));
            const span = Math.max(...yearNums) - Math.min(...yearNums);
            if (span > 0 && span <= 40) return span;
        }
        return 0;
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = ResumeParser;
}